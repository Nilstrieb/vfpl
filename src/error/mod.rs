mod display;
mod random;
mod span;

pub use display::display_error;
pub use random::random_ident;
pub use span::Span;

/// A trait for any error emitted during process
pub trait CompilerError {
    /// Return the span where the error occurred, for nicer error reporting. If no span is found,
    /// use [Span::dummy].
    fn span(&self) -> Span;

    /// The main message of the compiler error. It should be polite and not too long, but doesn't
    /// need extra verbosity.
    ///
    /// This needs to be a full sentence.
    ///
    /// # Examples
    ///
    /// `You have tried to add a String to a function.`
    fn message(&self) -> String;

    /// An additional note to give more context to the message. It can be long, but also needs to be
    /// polite.
    ///
    /// This needs to be a full sentence.
    ///
    /// # Examples
    /// `Due to the constraints of our spacetime, we have not found a sensible
    /// way to add a String to a function.`
    fn note(&self) -> Option<String> {
        None
    }

    /// A suggestion about what the programmer could do to resolve the error. It must be polite,
    /// and is allowed to be very wrong and not what the programmer intended, but it should fix the problem.
    ///
    /// This needs to be appendable to the phrase `You could `
    ///
    /// # Example
    /// `add another String to the String, for example "Function".`
    fn suggestion(&self) -> Option<String> {
        None
    }
}
