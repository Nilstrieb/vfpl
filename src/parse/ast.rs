use crate::error::Span;
use std::fmt::{Display, Formatter, Write};

type Ident = String;

pub type Program = Body;

#[derive(Debug, Clone, PartialEq)]
pub struct Body {
    pub span: Span,
    pub stmts: Vec<Stmt>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct TypedIdent {
    pub span: Span,
    pub name: Ident,
    pub ty: Ty,
}

#[derive(Debug, Clone, PartialEq)]
pub enum Stmt {
    VarInit(VarInit),
    VarSet(VarSet),
    Add(ArithmeticOp),
    Sub(ArithmeticOp),
    Mul(ArithmeticOp),
    Div(ArithmeticOp),
    Mod(ArithmeticOp),
    If(If),
    While(While),
    FnDecl(FnDecl),
    Break(Break),
    Return(Return),
    Terminate(Terminate),
    Expr(Expr),
}

impl Stmt {
    pub fn span(&self) -> Span {
        match self {
            Stmt::VarInit(inner) => inner.span,
            Stmt::VarSet(inner) => inner.span,
            Stmt::Add(inner)
            | Stmt::Sub(inner)
            | Stmt::Mul(inner)
            | Stmt::Div(inner)
            | Stmt::Mod(inner) => inner.span,
            Stmt::If(inner) => inner.span,
            Stmt::While(inner) => inner.span,
            Stmt::FnDecl(inner) => inner.span,
            Stmt::Break(inner) => inner.span,
            Stmt::Return(inner) => inner.span,
            Stmt::Terminate(inner) => inner.span,
            Stmt::Expr(expr) => expr.span(),
        }
    }
}

#[derive(Debug, Clone, PartialEq)]
pub struct VarInit {
    pub span: Span,
    pub name: TypedIdent,
    pub init: Expr,
}

#[derive(Debug, Clone, PartialEq)]
pub struct VarSet {
    pub span: Span,
    pub name: Ident,
    pub expr: Expr,
}

#[derive(Debug, Clone, PartialEq)]
pub struct ArithmeticOp {
    pub span: Span,
    pub expr: Expr,
    pub var: Ident,
    pub kind: ArithmeticOpKind,
}

#[derive(Debug, Clone, PartialEq)]
pub enum ArithmeticOpKind {
    Add,
    Sub,
    Mul,
    Div,
    Mod,
}

#[derive(Debug, Clone, PartialEq)]
pub struct If {
    pub span: Span,
    pub if_part: IfPart,
}

#[derive(Debug, Clone, PartialEq)]
pub struct IfPart {
    pub span: Span,
    pub cond: Box<Expr>,
    pub body: Body,
    pub else_part: Option<Else>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct Else {
    pub span: Span,
    pub kind: ElseKind,
}

#[derive(Debug, Clone, PartialEq)]
pub enum ElseKind {
    ElseIf(Box<IfPart>),
    Else(Body),
}

impl ElseKind {
    pub fn span(&self) -> Span {
        match self {
            ElseKind::Else(body) => body.span,
            ElseKind::ElseIf(if_) => if_.span,
        }
    }
}

#[derive(Debug, Clone, PartialEq)]
pub struct While {
    pub span: Span,
    pub cond: Expr,
    pub body: Body,
}

#[derive(Debug, Clone, PartialEq)]
pub struct Break {
    pub span: Span,
}

#[derive(Debug, Clone, PartialEq)]
pub struct FnDecl {
    pub span: Span,
    pub name: Ident,
    pub params: FnParams,
    pub fn_return: FnReturn,
    pub body: Body,
}

#[derive(Debug, Clone, PartialEq)]
pub struct FnParams {
    pub span: Span,
    pub params: Vec<TypedIdent>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct FnReturn {
    pub span: Span,
    pub ty: Ty,
}

#[derive(Debug, Clone, PartialEq)]
pub struct Return {
    pub span: Span,
    pub expr: Expr,
}

#[derive(Debug, Clone, PartialEq)]
pub struct Terminate {
    pub span: Span,
}

#[derive(Debug, Clone, PartialEq)]
pub struct Ty {
    pub span: Span,
    pub kind: TyKind,
}

#[derive(Debug, Clone, PartialEq)]
pub enum TyKind {
    Any,
    Name(Ident),
    Integer,
    Float,
    Boolean,
    String,
    Absent,
    Null,
    NoValue,
    Undefined,
}

#[derive(Debug, Clone, PartialEq)]
pub enum Expr {
    Literal(Literal),
    Comparison(Comparison),
    Call(Call),
}

impl Expr {
    pub fn span(&self) -> Span {
        match self {
            Expr::Literal(lit) => lit.span,
            Expr::Comparison(comp) => comp.span,
            Expr::Call(call) => call.span,
        }
    }
}

#[derive(Debug, Clone, PartialEq)]
pub struct Literal {
    pub span: Span,
    pub kind: LiteralKind,
}

#[derive(Debug, Clone, PartialEq)]
pub enum LiteralKind {
    Absent,
    Null,
    NoValue,
    Undefined,
    String(String),
    Int(i64),
    Float(f64),
    True,
    False,
    Ident(Ident),
}

#[derive(Debug, Clone, PartialEq)]
pub struct Comparison {
    pub span: Span,
    pub lhs: Box<Expr>,
    pub rhs: Box<Expr>,
    pub kind: ComparisonKind,
}

#[derive(Debug, Clone, PartialEq)]
pub enum ComparisonKind {
    NotEq,
    Eq,
    Greater,
    Less,
    GreaterEq,
    LessEq,
}

#[derive(Debug, Clone, PartialEq)]
pub struct Call {
    pub span: Span,
    pub fn_name: Ident,
    pub args: CallArgs,
}

#[derive(Debug, Clone, PartialEq)]
pub struct CallArgs {
    pub span: Span,
    pub args: Vec<CallArg>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct CallArg {
    pub span: Span,
    pub expr: Expr,
    pub name: Ident,
}

impl Display for ComparisonKind {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        match self {
            ComparisonKind::NotEq => f.write_str("!="),
            ComparisonKind::Eq => f.write_str("=="),
            ComparisonKind::Greater => f.write_char('>'),
            ComparisonKind::Less => f.write_char('<'),
            ComparisonKind::GreaterEq => f.write_str(">="),
            ComparisonKind::LessEq => f.write_str("<="),
        }
    }
}
