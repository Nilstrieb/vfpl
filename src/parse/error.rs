use crate::error::{CompilerError, Span};
use crate::VfplError;

#[derive(Debug)]
pub(super) struct ParseError {
    span: Span,
    message: String,
    note: Option<String>,
    suggestion: Option<String>,
}

impl ParseError {
    pub(super) fn simple(span: Span, message: String) -> Self {
        Self {
            span,
            message,
            note: None,
            suggestion: None,
        }
    }

    pub(super) fn full(span: Span, message: String, note: String, suggestion: String) -> Self {
        Self {
            span,
            message,
            note: Some(note),
            suggestion: Some(suggestion),
        }
    }
}

impl CompilerError for ParseError {
    fn span(&self) -> Span {
        self.span
    }

    fn message(&self) -> String {
        self.message.clone()
    }

    fn note(&self) -> Option<String> {
        self.note.clone()
    }

    fn suggestion(&self) -> Option<String> {
        self.suggestion.clone()
    }
}

impl From<ParseError> for VfplError {
    fn from(error: ParseError) -> Self {
        Self {
            span: error.span(),
            message: error.message(),
            note: error.note(),
            suggestion: error.suggestion(),
        }
    }
}
