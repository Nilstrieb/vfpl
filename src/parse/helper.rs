use crate::error::Span;
use crate::lexer::tokens::{Token, TokenType};
use crate::parse::{ParseError, ParseResult, Parser};

impl Parser {
    const MAX_DEPTH: usize = 500;

    pub(super) fn parse_rule<F, R>(&mut self, f: F) -> ParseResult<R>
    where
        F: FnOnce(&mut Self) -> ParseResult<R>,
    {
        self.enter_parse_rule()?;
        let result = f(self);
        self.leave_parse_rule();
        result
    }

    pub(super) fn next(&mut self) -> ParseResult<Token> {
        self.tokens.next().ok_or_else(|| ParseError {
            span: Span::dummy(),
            message: "reached end of file".to_string(),
        })
    }

    pub(super) fn peek_kind(&mut self) -> ParseResult<&TokenType> {
        self.tokens
            .peek()
            .map(|token| &token.kind)
            .ok_or_else(|| ParseError {
                span: Span::dummy(),
                message: "reached end of file".to_string(),
            })
    }

    pub(super) fn try_consume_kind(
        &mut self,
        expected_kind: TokenType,
    ) -> ParseResult<Option<Token>> {
        if self.peek_kind()? == &expected_kind {
            Ok(Some(self.next()?))
        } else {
            Ok(None)
        }
    }

    pub(super) fn expect_kind(&mut self, expected_kind: TokenType) -> ParseResult<Span> {
        let next = self.next()?;
        if next.kind == expected_kind {
            Ok(next.span)
        } else {
            Err(ParseError {
                span: next.span,
                message: format!("expected {}, found {}", expected_kind, next.kind),
            })
        }
    }

    pub(super) fn expect_kinds<const N: usize>(
        &mut self,
        expected_kinds: [TokenType; N],
    ) -> ParseResult<()> {
        for kind in expected_kinds {
            self.expect_kind(kind)?;
        }
        Ok(())
    }

    fn enter_parse_rule(&mut self) -> ParseResult<()> {
        self.depth += 1;
        if self.depth > Self::MAX_DEPTH {
            Err(ParseError {
                span: self
                    .tokens
                    .peek()
                    .map(|token| token.span)
                    .unwrap_or_else(Span::dummy),
                message: "Nesting too deep".to_string(),
            })
        } else {
            Ok(())
        }
    }

    fn leave_parse_rule(&mut self) {
        self.depth -= 1;
    }
}
