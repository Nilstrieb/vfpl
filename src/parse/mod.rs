#![allow(dead_code)]
use crate::error::Span;
use crate::lexer::tokens::Token;
use crate::parse::ast::{Body, Program, Stmt, TypedIdent};
use std::fmt::{Display, Formatter};
use std::iter::Peekable;
use std::vec::IntoIter;

pub mod ast;

type ParseResult<T> = Result<T, ParseError>;

type Todo = ();

#[derive(Debug)]
struct Parser {
    tokens: Peekable<IntoIter<Token>>,
    depth: usize,
}

pub fn parse(tokens: IntoIter<Token>) -> ParseResult<Program> {
    let mut parser = Parser {
        tokens: tokens.peekable(),
        depth: 0,
    };
    parser.program()
}

impl Parser {
    const MAX_DEPTH: usize = 500;

    fn enter_parse_rule(&mut self) -> ParseResult<()> {
        self.depth += 1;
        if self.depth > Self::MAX_DEPTH {
            Err(ParseError {
                span: Span::dummy(),
                message: "Nesting too deep".to_string(),
            })
        } else {
            Ok(())
        }
    }

    fn leave_parse_rule(&mut self) {
        self.depth -= 1;
    }

    fn next(&mut self) -> ParseResult<Token> {
        self.tokens.next().ok_or_else(|| ParseError {
            span: Span::dummy(),
            message: "reached end of file".to_string(),
        })
    }

    fn peek(&mut self) -> ParseResult<&Token> {
        self.tokens.peek().ok_or_else(|| ParseError {
            span: Span::dummy(),
            message: "reached end of file".to_string(),
        })
    }

    ////// rules

    fn program(&mut self) -> ParseResult<Program> {
        self.enter_parse_rule()?;
        self.leave_parse_rule();
        todo!()
    }
    fn body(&mut self) -> ParseResult<Body> {
        self.enter_parse_rule()?;
        self.leave_parse_rule();
        todo!()
    }
    fn typed_ident(&mut self) -> ParseResult<TypedIdent> {
        self.enter_parse_rule()?;
        self.leave_parse_rule();
        todo!()
    }
    fn stmt(&mut self) -> ParseResult<Stmt> {
        self.enter_parse_rule()?;
        self.leave_parse_rule();
        todo!()
    }
    fn var_init(&mut self) -> ParseResult<Todo> {
        self.enter_parse_rule()?;
        self.leave_parse_rule();
        todo!()
    }
    fn var_set(&mut self) -> ParseResult<Todo> {
        self.enter_parse_rule()?;
        self.leave_parse_rule();
        todo!()
    }
    fn add(&mut self) -> ParseResult<Todo> {
        self.enter_parse_rule()?;
        self.leave_parse_rule();
        todo!()
    }
    fn subtract(&mut self) -> ParseResult<Todo> {
        self.enter_parse_rule()?;
        self.leave_parse_rule();
        todo!()
    }
    fn multiply(&mut self) -> ParseResult<Todo> {
        self.enter_parse_rule()?;
        self.leave_parse_rule();
        todo!()
    }
    fn divide(&mut self) -> ParseResult<Todo> {
        self.enter_parse_rule()?;
        self.leave_parse_rule();
        todo!()
    }
    fn modulo(&mut self) -> ParseResult<Todo> {
        self.enter_parse_rule()?;
        self.leave_parse_rule();
        todo!()
    }
    fn if_stmt(&mut self) -> ParseResult<Todo> {
        self.enter_parse_rule()?;
        self.leave_parse_rule();
        todo!()
    }
    fn if_part(&mut self) -> ParseResult<Todo> {
        self.enter_parse_rule()?;
        self.leave_parse_rule();
        todo!()
    }
    fn else_stmt(&mut self) -> ParseResult<Todo> {
        self.enter_parse_rule()?;
        self.leave_parse_rule();
        todo!()
    }
    fn while_stmt(&mut self) -> ParseResult<Todo> {
        self.enter_parse_rule()?;
        self.leave_parse_rule();
        todo!()
    }
    fn break_stmt(&mut self) -> ParseResult<Todo> {
        self.enter_parse_rule()?;
        self.leave_parse_rule();
        todo!()
    }
    fn fn_decl(&mut self) -> ParseResult<Todo> {
        self.enter_parse_rule()?;
        self.leave_parse_rule();
        todo!()
    }
    fn params(&mut self) -> ParseResult<Todo> {
        self.enter_parse_rule()?;
        self.leave_parse_rule();
        todo!()
    }
    fn no_params(&mut self) -> ParseResult<Todo> {
        self.enter_parse_rule()?;
        self.leave_parse_rule();
        todo!()
    }
    fn single_param(&mut self) -> ParseResult<Todo> {
        self.enter_parse_rule()?;
        self.leave_parse_rule();
        todo!()
    }
    fn multi_param(&mut self) -> ParseResult<Todo> {
        self.enter_parse_rule()?;
        self.leave_parse_rule();
        todo!()
    }
    fn fn_return(&mut self) -> ParseResult<Todo> {
        self.enter_parse_rule()?;
        self.leave_parse_rule();
        todo!()
    }
    fn return_stmt(&mut self) -> ParseResult<Todo> {
        self.enter_parse_rule()?;
        self.leave_parse_rule();
        todo!()
    }
    fn terminate(&mut self) -> ParseResult<Todo> {
        self.enter_parse_rule()?;
        self.leave_parse_rule();
        todo!()
    }
    fn ty(&mut self) -> ParseResult<Todo> {
        self.enter_parse_rule()?;
        self.leave_parse_rule();
        todo!()
    }
    fn nullable(&mut self) -> ParseResult<Todo> {
        self.enter_parse_rule()?;
        self.leave_parse_rule();
        todo!()
    }
    fn literal(&mut self) -> ParseResult<Todo> {
        self.enter_parse_rule()?;
        self.leave_parse_rule();
        todo!()
    }
    fn number(&mut self) -> ParseResult<Todo> {
        self.enter_parse_rule()?;
        self.leave_parse_rule();
        todo!()
    }
    fn int(&mut self) -> ParseResult<Todo> {
        self.enter_parse_rule()?;
        self.leave_parse_rule();
        todo!()
    }
    fn float(&mut self) -> ParseResult<Todo> {
        self.enter_parse_rule()?;
        self.leave_parse_rule();
        todo!()
    }
    fn expr(&mut self) -> ParseResult<Todo> {
        self.enter_parse_rule()?;
        self.leave_parse_rule();
        todo!()
    }
    fn comparison(&mut self) -> ParseResult<Todo> {
        self.enter_parse_rule()?;
        self.leave_parse_rule();
        todo!()
    }
    fn primary_expr(&mut self) -> ParseResult<Todo> {
        self.enter_parse_rule()?;
        self.leave_parse_rule();
        todo!()
    }
    fn call(&mut self) -> ParseResult<Todo> {
        self.enter_parse_rule()?;
        self.leave_parse_rule();
        todo!()
    }
    fn call_args(&mut self) -> ParseResult<Todo> {
        self.enter_parse_rule()?;
        self.leave_parse_rule();
        todo!()
    }
    fn no_arg(&mut self) -> ParseResult<Todo> {
        self.enter_parse_rule()?;
        self.leave_parse_rule();
        todo!()
    }
    fn single_arg(&mut self) -> ParseResult<Todo> {
        self.enter_parse_rule()?;
        self.leave_parse_rule();
        todo!()
    }
    fn multi_arg(&mut self) -> ParseResult<Todo> {
        self.enter_parse_rule()?;
        self.leave_parse_rule();
        todo!()
    }
}

#[derive(Debug)]
pub struct ParseError {
    span: Span,
    message: String,
}

impl Display for ParseError {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        f.write_str(&self.message)
    }
}

impl std::error::Error for ParseError {}
