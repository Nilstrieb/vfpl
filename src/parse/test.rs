//// helpers

use crate::error::Span;
use crate::lexer::tokens::{
    Token,
    TokenKind::{self, *},
};
use crate::parse::Parser;

fn token(kind: TokenKind) -> Token {
    Token {
        span: Span::dummy(),
        kind,
    }
}

fn parse<T, F, R>(tokens: T, parse_rule_fn: F) -> R
where
    T: Into<Vec<Token>>,
    F: FnOnce(&mut Parser) -> R,
{
    let vec = tokens.into();
    let mut parser = Parser::new(vec.into_iter());
    parse_rule_fn(&mut parser)
}

//// tests

#[test]
fn it_works() {
    let snap = "test";
    insta::assert_debug_snapshot!(snap);
}

#[test]
fn single_string_literal() {
    let tokens = [String("string".to_string())].map(token);
    let parsed = parse(tokens, Parser::literal);
    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn different_literals() {
    let literals = [
        [String("test".to_string())].map(token),
        [Int(425)].map(token),
        [Float(2465.67)].map(token),
        [Absent].map(token),
        [NoValue].map(token),
        [Null].map(token),
        [Undefined].map(token),
        [True].map(token),
        [False].map(token),
    ];

    let parsed = literals.map(|tokens| parse(tokens, Parser::literal));

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn ident_ty() {
    let tokens = [Ident("Test".to_string())].map(token);
    let parsed = parse(tokens, Parser::ty);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn nullable_ty() {
    let tys = [
        [Absent].map(token),
        [NoValue].map(token),
        [Null].map(token),
        [Undefined].map(token),
    ];
    let parsed = tys.map(|tokens| parse(tokens, Parser::ty));

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn other_tys() {
    let tys = [
        [Ident("Boolean".to_string())].map(token),
        [Ident("Integer".to_string())].map(token),
        [Ident("String".to_string())].map(token),
        [Ident("Float".to_string())].map(token),
    ];
    let parsed = tys.map(|tokens| parse(tokens, Parser::ty));

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn literal_call_expr() {
    let tokens = [String("string".to_string())].map(token);
    let parsed = parse(tokens, Parser::call_expr);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn literal_expr() {
    let tokens = [String("string".to_string())].map(token);
    let parsed = parse(tokens, Parser::expr);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn ident_literal() {
    let tokens = [Ident("uwu".to_string())].map(token);
    let parsed = parse(tokens, Parser::expr);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn not_equal_expr() {
    let tokens = [Absent, Does, Not, Have, The, Value, Null].map(token);
    let parsed = parse(tokens, Parser::expr);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn equal_expr() {
    let tokens = [Absent, Has, The, Value, Absent].map(token);
    let parsed = parse(tokens, Parser::expr);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn greater_than_expr() {
    let tokens = [Absent, Is, Greater, Than, Undefined].map(token);
    let parsed = parse(tokens, Parser::expr);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn less_than_expr() {
    let tokens = [Absent, Is, Less, Than, Undefined].map(token);
    let parsed = parse(tokens, Parser::expr);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn greater_equal_than_expr() {
    let tokens = [Absent, Is, Greater, Or, Equal, Than, Undefined].map(token);
    let parsed = parse(tokens, Parser::expr);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn less_equal_than_expr() {
    let tokens = [Absent, Is, Less, Or, Equal, Than, Undefined].map(token);
    let parsed = parse(tokens, Parser::expr);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn typed_ident_absent() {
    let tokens = [Ident("name".to_string()), As, Absent].map(token);
    let parsed = parse(tokens, Parser::typed_ident);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn init_variable_string() {
    let tokens = [
        Initialize,
        Variable,
        Ident("name".to_string()),
        As,
        Ident("string".to_string()),
        With,
        The,
        Value,
        Of,
        String("Ferris".to_string()),
    ]
    .map(token);
    let parsed = parse(tokens, Parser::var_init);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn set_variable_string() {
    let tokens = [
        Set,
        The,
        Variable,
        Ident("name".to_string()),
        To,
        The,
        Value,
        Of,
        String("Ferris".to_string()),
    ]
    .map(token);
    let parsed = parse(tokens, Parser::var_set);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn add_number() {
    let tokens = [Add, Int(0), To, Ident("counter".to_string())].map(token);
    let parsed = parse(tokens, Parser::add);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn subtract_number() {
    let tokens = [Sub, Int(1), From, Ident("counter".to_string())].map(token);
    let parsed = parse(tokens, Parser::subtract);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn divide_number() {
    let tokens = [Div, Ident("counter".to_string()), By, Int(2)].map(token);
    let parsed = parse(tokens, Parser::divide);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn multiply_number() {
    let tokens = [Mul, Ident("counter".to_string()), With, Int(3)].map(token);
    let parsed = parse(tokens, Parser::multiply);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn mod_number() {
    let tokens = [Take, Ident("counter".to_string()), Mod, Int(4)].map(token);
    let parsed = parse(tokens, Parser::modulo);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn terminate() {
    let tokens = [Go, To, Sleep].map(token);
    let parsed = parse(tokens, Parser::terminate);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn call_no_args() {
    let tokens = [Call, Ident("run".to_string()), With, No, Arguments].map(token);
    let parsed = parse(tokens, Parser::call);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn call_single_arg() {
    let tokens = [
        Call,
        Ident("print".to_string()),
        With,
        The,
        Argument,
        Int(0),
        As,
        Ident("printable".to_string()),
    ]
    .map(token);
    let parsed = parse(tokens, Parser::call);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn call_two_args() {
    let tokens = [
        Call,
        Ident("add".to_string()),
        With,
        The,
        Arguments,
        Int(2),
        As,
        Ident("a".to_string()),
        And,
        Int(3),
        As,
        Ident("b".to_string()),
    ]
    .map(token);
    let parsed = parse(tokens, Parser::call);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn call_three_args() {
    let tokens = [
        Call,
        Ident("ternary".to_string()),
        With,
        The,
        Arguments,
        True,
        As,
        Ident("cond".to_string()),
        Comma,
        Int(0),
        As,
        Ident("then".to_string()),
        And,
        Int(3),
        As,
        Ident("else".to_string()),
    ]
    .map(token);
    let parsed = parse(tokens, Parser::call);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn add_stmt() {
    let tokens = [Please, Add, Int(0), To, Ident("A".to_string()), Dot].map(token);
    let parsed = parse(tokens, Parser::stmt);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn if_empty_body_true_cond() {
    let tokens = [Check, Whether, True, Comma, Then, Do, Please, End, Check].map(token);
    let parsed = parse(tokens, Parser::if_stmt);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn if_empty_body_complex_cond() {
    let tokens = [
        Check, Whether, True, Has, The, Value, Absent, Comma, Then, Do, Please, End, Check,
    ]
    .map(token);
    let parsed = parse(tokens, Parser::if_stmt);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn if_single_stmt_body() {
    let tokens = [
        Check, Whether, True, Has, The, Value, Absent, Comma, Then, Do, Please, Go, To, Sleep, Dot,
        Please, End, Check,
    ]
    .map(token);
    let parsed = parse(tokens, Parser::if_stmt);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn if_multi_stmt_body() {
    let tokens = [
        Check,
        Whether,
        True,
        Has,
        The,
        Value,
        Absent,
        Comma,
        Then,
        Do,
        Please,
        Go,
        To,
        Sleep,
        Dot,
        Please,
        Add,
        Int(5),
        To,
        Ident("A".to_string()),
        Dot,
        Please,
        End,
        Check,
    ]
    .map(token);
    let parsed = parse(tokens, Parser::if_stmt);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn return_number() {
    let tokens = [Return, Int(9), From, The, Function].map(token);
    let parsed = parse(tokens, Parser::return_stmt);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn return_stmt() {
    let tokens = [Please, Return, Int(9), From, The, Function, Dot].map(token);
    let parsed = parse(tokens, Parser::stmt);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn r#break() {
    let tokens = [Break, Out, Of, This, While].map(token);
    let parsed = parse(tokens, Parser::break_stmt);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn break_stmt() {
    let tokens = [Please, Break, Out, Of, This, While, Dot].map(token);
    let parsed = parse(tokens, Parser::stmt);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn while_loop_no_body() {
    let tokens = [Repeat, While, True, Do, Please, End, While].map(token);
    let parsed = parse(tokens, Parser::while_stmt);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn while_stmt_with_body() {
    let tokens = [
        Please,
        Repeat,
        While,
        Absent,
        Has,
        The,
        Value,
        False,
        Do,
        Please,
        Add,
        Int(9),
        To,
        Ident("A".to_string()),
        Dot,
        Please,
        End,
        While,
        Dot,
    ]
    .map(token);
    let parsed = parse(tokens, Parser::stmt);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn return_ty_absent() {
    let tokens = [That, Returns, Absent].map(token);
    let parsed = parse(tokens, Parser::fn_return);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn fn_decl_no_params_empty_body() {
    let tokens = [
        Create,
        Function,
        Ident("void".to_string()),
        With,
        No,
        Parameters,
        That,
        Returns,
        Absent,
        Please,
        End,
        Function,
        Ident("void".to_string()),
    ]
    .map(token);
    let parsed = parse(tokens, Parser::fn_decl);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn fn_decl_single_param_emtpy_body() {
    let tokens = [
        Create,
        Function,
        Ident("print".to_string()),
        With,
        The,
        Parameter,
        Ident("_".to_string()),
        As,
        Absent,
        That,
        Returns,
        Absent,
        Please,
        End,
        Function,
        Ident("print".to_string()),
    ]
    .map(token);
    let parsed = parse(tokens, Parser::fn_decl);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn fn_decl_two_param_emtpy_body() {
    let tokens = [
        Create,
        Function,
        Ident("add".to_string()),
        With,
        The,
        Parameters,
        Ident("_".to_string()),
        As,
        Absent,
        And,
        Ident("_hi".to_string()),
        As,
        Null,
        That,
        Returns,
        Absent,
        Please,
        End,
        Function,
        Ident("add".to_string()),
    ]
    .map(token);
    let parsed = parse(tokens, Parser::fn_decl);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn fn_decl_three_param_with_body() {
    let tokens = [
        Create,
        Function,
        Ident("add".to_string()),
        With,
        The,
        Parameters,
        Ident("_".to_string()),
        As,
        Absent,
        Comma,
        Ident("a".to_string()),
        As,
        NoValue,
        And,
        Ident("_hi".to_string()),
        As,
        Null,
        That,
        Returns,
        Absent,
        Please,
        Add,
        Int(5),
        To,
        Ident("a".to_string()),
        Dot,
        Please,
        End,
        Function,
        Ident("add".to_string()),
    ]
    .map(token);
    let parsed = parse(tokens, Parser::fn_decl);

    insta::assert_debug_snapshot!(parsed);
}
