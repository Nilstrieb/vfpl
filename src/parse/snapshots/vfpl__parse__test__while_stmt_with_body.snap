---
source: src/parse/test.rs
expression: parsed

---
Ok(
    While(
        While {
            span: Span {
                start: 0,
                end: 0,
            },
            cond: Comparison(
                Comparison {
                    span: Span {
                        start: 0,
                        end: 0,
                    },
                    lhs: Literal(
                        Literal {
                            span: Span {
                                start: 0,
                                end: 0,
                            },
                            kind: Absent,
                        },
                    ),
                    rhs: Literal(
                        Literal {
                            span: Span {
                                start: 0,
                                end: 0,
                            },
                            kind: False,
                        },
                    ),
                    kind: Eq,
                },
            ),
            body: Body {
                span: Span {
                    start: 0,
                    end: 0,
                },
                stmts: [
                    Add(
                        ArithmeticOp {
                            span: Span {
                                start: 0,
                                end: 0,
                            },
                            expr: Literal(
                                Literal {
                                    span: Span {
                                        start: 0,
                                        end: 0,
                                    },
                                    kind: Int(
                                        9,
                                    ),
                                },
                            ),
                            var: "A",
                            kind: Add,
                        },
                    ),
                ],
            },
        },
    ),
)
